narwhals/_arrow/dataframe.py

```diff
@@ -743,3 +744,78 @@ def unpivot(
         )
         # TODO(Unassigned): Even with promote_options="permissive", pyarrow does not
         # upcast numeric to non-numeric (e.g. string) datatypes
+
+    def explode(self: Self, columns: str | Sequence[str], *more_columns: str) -> Self:
```

Comment:
pyarrow has two paths:

- if nulls or empty lists are not present, then it is enough to:
  1. make sure element counts are same
  2. explode each array individually
- if nulls or empty lists are present, then these are ignore by `pc.list_parent_indices` and `pc.list_flatten`, which _is_ a problem. This implementation falls back to a python list both to flatten the array(s) and to create the corresponding indices .

After flattening, a new table is created by `take`-ing the indices of the non-flattened arrays and the flattened arrays.